File: packages/SystemUI/src/com/benzorom/systemui/keyguard/KeyguardSliceProviderGoogle.java
package com.benzorom.systemui.keyguard;

import android.app.PendingIntent;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.BlurMaskFilter;
import android.graphics.BlurMaskFilter.Blur;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff.Mode;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Trace;
import android.text.TextUtils;
import android.util.Log;

import androidx.core.graphics.drawable.IconCompat;
import androidx.slice.Slice;
import androidx.slice.builders.ListBuilder;
import androidx.slice.builders.ListBuilder.HeaderBuilder;
import androidx.slice.builders.ListBuilder.RowBuilder;
import androidx.slice.builders.SliceAction;

import com.android.systemui.R;
import com.android.systemui.SystemUIFactory;
import com.android.systemui.keyguard.KeyguardSliceProvider;
import com.google.android.systemui.smartspace.SmartSpaceCard;
import com.google.android.systemui.smartspace.SmartSpaceController;
import com.google.android.systemui.smartspace.SmartSpaceData;
import com.google.android.systemui.smartspace.SmartSpaceUpdateListener;

import java.lang.ref.WeakReference;

import javax.inject.Inject;

public class KeyguardSliceProviderGoogle extends KeyguardSliceProvider
        implements SmartSpaceUpdateListener {

    private static final String TAG = "KeyguardSliceProvider";
    private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
    private final Uri mCalendarUri = Uri.parse("content://com.android.systemui.keyguard/smartSpace/calendar");
    private final Uri mWeatherUri = Uri.parse("content://com.android.systemui.keyguard/smartSpace/weather");

    @Inject
    public SmartSpaceController mSmartSpaceController;
    private SmartSpaceData mSmartSpaceData;
    private boolean mHideSensitiveContent;
    private boolean mHideWorkContent = true;

    private static class AddShadowTask extends AsyncTask<Bitmap, Void, Bitmap> {
        private final float mBlurRadius;
        private final WeakReference<KeyguardSliceProviderGoogle> mProviderReference;
        private final SmartSpaceCard mWeatherCard;

        AddShadowTask(
                KeyguardSliceProviderGoogle keyguardSliceProviderGoogle,
                SmartSpaceCard smartSpaceCard
        ) {
            mProviderReference = new WeakReference<>(keyguardSliceProviderGoogle);
            mWeatherCard = smartSpaceCard;
            mBlurRadius = keyguardSliceProviderGoogle.getContext().getResources().getDimension(
                    R.dimen.smartspace_icon_shadow
            );
        }

        @Override
        protected Bitmap doInBackground(Bitmap... bitmapArr) {
            return applyShadow(bitmapArr[0]);
        }

        @Override
        protected void onPostExecute(Bitmap bitmap) {
            KeyguardSliceProviderGoogle keyguardSliceProviderGoogle;
            synchronized (this) {
                mWeatherCard.setIcon(bitmap);
                keyguardSliceProviderGoogle = (KeyguardSliceProviderGoogle) mProviderReference.get();
            }
            if (keyguardSliceProviderGoogle != null) {
                keyguardSliceProviderGoogle.notifyChange();
            }
        }

        private Bitmap applyShadow(Bitmap bitmap) {
            BlurMaskFilter blurMaskFilter = new BlurMaskFilter(mBlurRadius, Blur.NORMAL);
            Paint paint = new Paint();
            paint.setMaskFilter(blurMaskFilter);
            int[] iArr = new int[2];
            Bitmap extractAlpha = bitmap.extractAlpha(paint, iArr);
            Bitmap createBitmap = Bitmap.createBitmap(bitmap.getWidth(), bitmap.getHeight(), Config.ARGB_8888);
            Canvas canvas = new Canvas(createBitmap);
            Paint paint2 = new Paint();
            paint2.setAlpha(70);
            canvas.drawBitmap(extractAlpha, (float) iArr[0], ((float) iArr[1]) + (mBlurRadius / 2.0f), paint2);
            extractAlpha.recycle();
            paint2.setAlpha(255);
            canvas.drawBitmap(bitmap, 0.0f, 0.0f, paint2);
            return createBitmap;
        }
    }

    @Override
    public boolean onCreateSliceProvider() {
        boolean sliceProvider = super.onCreateSliceProvider();
        mSmartSpaceData = new SmartSpaceData();
        mSmartSpaceController.addListener(this);
        return sliceProvider;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mSmartSpaceController.removeListener(this);
    }

    @Override
    public Slice onBindSlice(Uri sliceUri) {
        Trace.beginSection("KeyguardSliceProviderGoogle#onBindSlice");
        ListBuilder sliceBuilder = new ListBuilder(getContext(), mSliceUri, -1);
        synchronized (this) {
            SmartSpaceCard currentCard = mSmartSpaceData.getCurrentCard();
            boolean hasAction = false;
            if (currentCard != null && !currentCard.isExpired() &&
                    !TextUtils.isEmpty(currentCard.getTitle())) {
                boolean isSensitive = currentCard.isSensitive();
                if (!isSensitive || isSensitive &&
                        (!mHideSensitiveContent && !currentCard.isWorkProfile()
                        || !mHideWorkContent && currentCard.isWorkProfile())) {
                    hasAction = true;
                }
            }
            if (hasAction) {
                Bitmap icon = currentCard.getIcon();
                SliceAction sliceAction = null;
                IconCompat createWithBitmap = icon == null ? null : IconCompat.createWithBitmap(icon);
                PendingIntent pendingIntent = currentCard.getPendingIntent();
                if (createWithBitmap != null || pendingIntent != null) {
                    sliceAction = SliceAction.create(pendingIntent, createWithBitmap, 1, currentCard.getTitle());
                }
                HeaderBuilder headerBuilder = new HeaderBuilder(mHeaderUri);
                headerBuilder.setTitle(currentCard.getFormattedTitle());
                if (sliceAction != null) {
                    headerBuilder.setPrimaryAction(sliceAction);
                }
                sliceBuilder.setHeader(headerBuilder);
                String subtitle = currentCard.getSubtitle();
                if (subtitle != null) {
                    RowBuilder calendarBuilder = new RowBuilder(mCalendarUri).setTitle(subtitle);
                    if (createWithBitmap != null) {
                        calendarBuilder.addEndItem(createWithBitmap, 1);
                    }
                    if (sliceAction != null) {
                        calendarBuilder.setPrimaryAction(sliceAction);
                    }
                    sliceBuilder.addRow(calendarBuilder);
                }
                addZenModeLocked(sliceBuilder);
                addPrimaryActionLocked(sliceBuilder);
                Trace.endSection();
                return sliceBuilder.build();
            }
            if (needsMediaLocked()) {
                addMediaLocked(sliceBuilder);
            } else {
                sliceBuilder.addRow(new RowBuilder(mDateUri).setTitle(getFormattedDateLocked()));
            }
            addWeather(sliceBuilder);
            addNextAlarmLocked(sliceBuilder);
            addZenModeLocked(sliceBuilder);
            addPrimaryActionLocked(sliceBuilder);
            Slice slice = sliceBuilder.build();
            if (DEBUG) Log.d(TAG, "Binding slice: " + slice);
            Trace.endSection();
            return slice;
        }
    }

    private void addWeather(ListBuilder sliceBuilder) {
        SmartSpaceCard weatherCard = mSmartSpaceData.getWeatherCard();
        if (weatherCard != null && !weatherCard.isExpired()) {
            RowBuilder weatherBuilder = new RowBuilder(mWeatherUri);
            weatherBuilder.setTitle(weatherCard.getTitle());
            Bitmap icon = weatherCard.getIcon();
            if (icon != null) {
                IconCompat createWithBitmap = IconCompat.createWithBitmap(icon);
                createWithBitmap.setTintMode(Mode.DST);
                weatherBuilder.addEndItem(createWithBitmap, 1);
            }
            sliceBuilder.addRow(weatherBuilder);
        }
    }

    @Override
    public void onSmartSpaceUpdated(SmartSpaceData smartSpaceData) {
        synchronized (this) {
            mSmartSpaceData = smartSpaceData;
        }
        SmartSpaceCard weatherCard = smartSpaceData.getWeatherCard();
        if (weatherCard == null || weatherCard.getIcon() == null || weatherCard.isIconProcessed()) {
            notifyChange();
            return;
        }
        weatherCard.setIconProcessed(true);
        new AddShadowTask(this, weatherCard).execute(new Bitmap[]{weatherCard.getIcon()});
    }

    @Override
    public void onSensitiveModeChanged(boolean hideSensitiveContent, boolean hideWorkContent) {
        synchronized (this) {
            boolean changed = false;
            if (mHideSensitiveContent != hideSensitiveContent) {
                mHideSensitiveContent = hideSensitiveContent;
                if (DEBUG) Log.d(TAG, "Public mode changed, hide data: " + hideSensitiveContent);
                changed = true;
            }
            if (mHideWorkContent != hideWorkContent) {
                mHideWorkContent = hideWorkContent;
                if (DEBUG) Log.d(TAG, "Public work mode changed, hide data: " + hideWorkContent);
                changed = true;
            }
            if (changed) {
                notifyChange();
            }
        }
    }

    @Override
    public void updateClockLocked() {
        notifyChange();
    }
}

File: packages/SystemUI/src/com/benzorom/systemui/keyguard/KeyguardSliceProviderGoogle.kt
/*
 * Copyright (C) 2022 Benzo Rom
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:Suppress("DEPRECATION")
package com.benzorom.systemui.keyguard

import android.annotation.NonNull
import android.annotation.Nullable
import android.graphics.*
import android.net.Uri
import android.os.AsyncTask
import android.os.Trace
import android.text.TextUtils
import android.util.Log
import androidx.core.graphics.drawable.IconCompat
import androidx.slice.Slice
import androidx.slice.builders.ListBuilder
import androidx.slice.builders.ListBuilder.HeaderBuilder
import androidx.slice.builders.SliceAction
import com.android.systemui.R
import com.android.systemui.keyguard.KeyguardSliceProvider
import com.google.android.systemui.smartspace.SmartSpaceCard
import com.google.android.systemui.smartspace.SmartSpaceController
import com.google.android.systemui.smartspace.SmartSpaceData
import com.google.android.systemui.smartspace.SmartSpaceUpdateListener
import java.lang.ref.WeakReference
import javax.inject.Inject

class KeyguardSliceProviderGoogle : KeyguardSliceProvider(), SmartSpaceUpdateListener {

    companion object {
        private const val logTag = "KeyguardSliceProvider"
        private val isDebug = Log.isLoggable(logTag, Log.DEBUG)
    }

    private val weatherUri = Uri.parse("content://com.android.systemui.keyguard/smartSpace/weather")
    private val calendarUri = Uri.parse("content://com.android.systemui.keyguard/smartSpace/calendar")
    @Inject lateinit var smartSpaceController: SmartSpaceController
    private var smartSpaceData: SmartSpaceData? = null
    private var hideSensitiveContent = false
    private var hideWorkContent = true

    override fun onCreateSliceProvider(): Boolean {
        val created = super.onCreateSliceProvider()
        smartSpaceData = SmartSpaceData()
        smartSpaceController.addListener(this)
        return created
    }

    override fun onDestroy() {
        super.onDestroy()
        smartSpaceController.removeListener(this)
    }

    override fun onBindSlice(sliceUri: Uri): Slice {
        Trace.beginSection("KeyguardSliceProviderGoogle#onBindSlice")
        var slice: Slice
        synchronized(this) {
            val sliceBuilder = ListBuilder(
                context!!,
                mSliceUri,
                ListBuilder.INFINITY
            )
            val card = smartSpaceData!!.currentCard
            var hasAction = false
            if (card != null && !card.isExpired && !TextUtils.isEmpty(card.title)) {
                if (!card.isSensitive || card.isSensitive &&
                    (!hideSensitiveContent && !card.isWorkProfile
                            || !hideWorkContent && card.isWorkProfile)
                ) hasAction = true
            }
            if (hasAction) {
                val icon = card!!.icon
                var action: SliceAction? = null
                val iconBitmap = if (icon != null) IconCompat.createWithBitmap(icon) else null
                val intent = card.pendingIntent
                if (!(iconBitmap == null || intent == null))
                    action = SliceAction.create(
                        intent,
                        iconBitmap,
                        ListBuilder.SMALL_IMAGE,
                        card.title
                    )
                val title = HeaderBuilder(mHeaderUri).setTitle(card.formattedTitle)
                if (action != null) title.setPrimaryAction(action)
                sliceBuilder.setHeader(title)
                val subtitle = card.subtitle
                if (subtitle != null) {
                    val calendar = ListBuilder.RowBuilder(calendarUri).setTitle(subtitle)
                    if (iconBitmap != null)
                        calendar.addEndItem(
                            iconBitmap,
                            ListBuilder.SMALL_IMAGE
                        )
                    if (action != null) calendar.setPrimaryAction(action)
                    sliceBuilder.addRow(calendar)
                }
                addZenModeLocked(sliceBuilder)
                addPrimaryActionLocked(sliceBuilder)
                slice = sliceBuilder.build()
                Trace.endSection()
                return slice
            }
            if (needsMediaLocked()) {
                addMediaLocked(sliceBuilder)
            } else {
                sliceBuilder.addRow(
                    ListBuilder.RowBuilder(mDateUri).setTitle(formattedDateLocked)
                )
            }
            addWeather(sliceBuilder)
            addNextAlarmLocked(sliceBuilder)
            addZenModeLocked(sliceBuilder)
            addPrimaryActionLocked(sliceBuilder)
            slice = sliceBuilder.build()
            if (isDebug) Log.d(logTag, "Binding slice: $slice")
            Trace.endSection()
            return slice
        }
    }

    private fun addWeather(sliceBuilder: ListBuilder) {
        val weatherCard = smartSpaceData!!.weatherCard
        if (weatherCard != null && !weatherCard.isExpired) {
            val title = ListBuilder.RowBuilder(weatherUri).setTitle(weatherCard.title)
            val icon = weatherCard.icon
            if (icon != null) {
                val iconBitmap = IconCompat.createWithBitmap(icon)
                iconBitmap.setTintMode(PorterDuff.Mode.DST)
                title.addEndItem(iconBitmap, ListBuilder.SMALL_IMAGE)
            }
            sliceBuilder.addRow(title)
        }
    }

    override fun onSensitiveModeChanged(hideSensitive: Boolean, hideWork: Boolean) {
        var notify = false
        synchronized(this) {
            if (hideSensitiveContent != hideSensitive) {
                hideSensitiveContent = hideSensitive
                notify = true
                if (isDebug) Log.d(logTag, "Public mode changed, hide data: $hideSensitive")
            }
            if (hideWorkContent != hideWork) {
                hideWorkContent = hideWork
                notify = true
                if (isDebug) Log.d(logTag, "Public work mode changed, hide data: $hideWork")
            }
        }
        if (notify) notifyChange()
    }

    override fun onSmartSpaceUpdated(data: SmartSpaceData) {
        synchronized(this) { smartSpaceData = data }
        val weatherCard = data.weatherCard
        if (weatherCard == null || weatherCard.icon == null || weatherCard.isIconProcessed) {
            notifyChange()
        } else {
            weatherCard.isIconProcessed = true
            AddShadowTask(this, weatherCard).execute(weatherCard.icon)
        }
    }

    override fun updateClockLocked() {
        notifyChange()
    }

    private class AddShadowTask(
        provider: KeyguardSliceProviderGoogle,
        card: SmartSpaceCard
    ) : AsyncTask<Bitmap?, Void?, Bitmap>() {
        private val blurRadius: Float
        private val providerReference: WeakReference<KeyguardSliceProviderGoogle>
        private val weatherCard: SmartSpaceCard

        private fun applyShadow(@NonNull result: Bitmap): Bitmap {
            val blur = BlurMaskFilter(blurRadius, BlurMaskFilter.Blur.NORMAL)
            val blurPaint = Paint()
            blurPaint.maskFilter = blur
            val offset = IntArray(2)
            val alpha = result.extractAlpha(blurPaint, offset)
            val bitmap = Bitmap.createBitmap(
                result.width,
                result.height,
                Bitmap.Config.ARGB_8888
            )
            val canvas = Canvas(bitmap)
            val paint = Paint()
            paint.alpha = 70
            canvas.drawBitmap(
                alpha,
                offset[0].toFloat(),
                offset[1].toFloat().plus(blurRadius / 2.0f),
                paint
            )
            alpha.recycle()
            paint.alpha = 255
            canvas.drawBitmap(
                result,
                0.0f,
                0.0f,
                paint
            )
            return bitmap
        }

        @Deprecated("Deprecated in Java")
        override fun doInBackground(@NonNull vararg bitmaps: Bitmap?): Bitmap {
            return applyShadow(bitmaps[0]!!)
        }

        @Deprecated("Deprecated in Java")
        override fun onPostExecute(@Nullable result: Bitmap) {
            var provider: KeyguardSliceProviderGoogle?
            synchronized(this) {
                weatherCard.icon = result
                provider = providerReference.get()
            }
            if (provider != null) {
                provider!!.notifyChange()
            }
        }

        init {
            blurRadius = provider.context!!.resources.getDimension(R.dimen.smartspace_icon_shadow)
            providerReference = WeakReference(provider)
            weatherCard = card
        }
    }
}
